code/game/machinery/washing_machine.dm

```diff
@@ -57,283 +118,307 @@
-			if(new_bandana_icon_state && new_bandana_name)
-				for(var/obj/item/clothing/mask/bandana/M in contents)
-					if(!M.dyeable)
-						continue
-					M.item_state = new_bandana_item_state
-					M.icon_state = new_bandana_icon_state
-					M.item_color = wash_color
-					M.name = new_bandana_name
-					M.desc = new_desc
-			if(new_sheet_icon_state && new_sheet_name)
-				for(var/obj/item/bedsheet/B in contents)
-					B.icon_state = new_sheet_icon_state
-					B.item_color = wash_color
-					B.name = new_sheet_name
-					B.desc = new_desc
-			if(new_softcap_icon_state && new_softcap_name)
-				for(var/obj/item/clothing/head/soft/H in contents)
-					if(!H.dyeable)
-						continue
-					H.icon_state = new_softcap_icon_state
-					H.item_color = wash_color
-					H.name = new_softcap_name
-					H.desc = new_desc
+			for(var/obj/item/Item in contents)
+				var/shouldbreak
+				if(istype(Item, /obj/item/clothing)) //if not dyeable, skip this item
+					for(var/T in typesof(/obj/item/clothing))
+						var/obj/item/clothing/X = new T
+						if(Item == X)
+							if(!X.dyeable)
+								shouldbreak = TRUE
+							break
+				if(shouldbreak)
+					continue
+				if(istype(Item, /obj/item/clothing))
+					dyethis(Item, wash_color, TRUE)
+				else
+					dyethis(Item, wash_color, FALSE)
 		QDEL_NULL(crayon)
 
-
-	if( locate(/mob,contents) )
-		state = 7
-		gibs_ready = 1
+	running = FALSE
+	if(!full)
+		bloody = FALSE
 	else
-		state = 4
+		playsound(src, 'sound/weapons/jug_filled_impact.ogg', 10)
 	update_icon()
+	sleep(5) // todo: replace this with a callback
+	if(stat & NOPOWER)
+		playsound(src, 'sound/machines/terminal_off.ogg', 45)
+	else
+		playsound(src, 'sound/machines/defib_success.ogg', 25)
 
-/obj/machinery/washing_machine/verb/climb_out()
-	set name = "Climb out"
-	set category = "Object"
-	set src in usr.loc
-
-	sleep(20)
-	if(state in list(1,3,6) )
-		usr.loc = src.loc
-
+/obj/machinery/washing_machine/proc/dyethis(var/obj/item/Item, wash_color, clothing = TRUE) // god this is nasty
+	// Things that are acceptable to dye
+	var/list/dyeablethings = list(/obj/item/clothing/under/plasmaman, /obj/item/clothing/head/helmet/space/plasmaman, /obj/item/clothing/under, /obj/item/clothing/gloves/color, /obj/item/clothing/shoes, /obj/item/clothing/mask/bandana, /obj/item/clothing/head/soft, /obj/item/clothing/head/beanie, /obj/item/clothing/accessory/scarf, /obj/item/bedsheet, /obj/item/storage/wallet/color, /obj/item/toy/carpplushie)
+	var/list/possible = list()
+	for(var/obj/item/Path in dyeablethings) // for every item in contents, check every it against every dyeablething, then check that for every subtype (god what the fuck)
+		if(istype(Item, Path))
+			for(var/T in typesof(Path))
+				var/obj/item/J = new T // for the love of god change the variable name
+				if(clothing)
+					var/obj/item/clothing/J2 = T // this one too
+					if(!initial(J2.dyeable))
+						continue
+				if((wash_color == J.item_color))
+					possible += J
+				qdel(J)
```

Comment:
```suggestion
	for(var/path in dyeablethings)
		if(istype(Item, path))
				if(clothing)
					var/obj/item/clothing/apparel = T // this one too
					if(!apparel.dyeable)
						continue
				if(wash_color == Item.item_color)
					possible += path
```
do not create new objects just to interface with that type's variables unless you have no choice, also you can't even interface with a specific types variable if you always new it as only a `obj/item`. As long as `item_color` and `dyeable` are base item object variables you don't care about their paths. You can then store the "good" objects by their paths which will cut down on the performance hog.